ignite/metrics/ssim.py

```diff
@@ -0,0 +1,144 @@
+    ):
+        if isinstance(kernel_size, int):
+            self.kernel_size = [kernel_size, kernel_size]
+        elif isinstance(kernel_size, Sequence):
+            self.kernel_size = kernel_size
+        else:
+            raise ValueError("Argument kernel_size should be either int or a sequence of int.")
+
+        if isinstance(sigma, float):
+            self.sigma = [sigma, sigma]
+        elif isinstance(sigma, Sequence):
+            self.sigma = sigma
+        else:
+            raise ValueError("Argument sigma should be either float or a sequence of float.")
+
+        if any(x % 2 == 0 or x <= 0 for x in self.kernel_size):
+            raise ValueError("Expected kernel_size to have odd positive number. Got {}.".format(kernel_size))
+
+        if any(y <= 0 for y in self.sigma):
+            raise ValueError("Expected sigma to have positive number. Got {}.".format(sigma))
+
+        self.data_range = data_range
+        self.k1 = k1
+        self.k2 = k2
+        self.kernel = self._gaussian_kernel(kernel_size=self.kernel_size, sigma=self.sigma)
+        super(SSIM, self).__init__(output_transform=output_transform)
+
+    @reinit__is_reduced
+    def reset(self) -> None:
+        self._sum_of_batchwise_ssim = 0.0
+        self._num_examples = 0
+
+    def _gaussian(self, kernel_size, sigma):
+        gauss = torch.arange(start=(1 - kernel_size) / 2, end=(1 + kernel_size) / 2, step=1, dtype=torch.float32)
+        gauss = torch.exp(-gauss.pow(2) / (2 * pow(sigma, 2)))
+        return (gauss / gauss.sum()).unsqueeze(dim=0)  # (1, kernel_size)
+
+    def _gaussian_kernel(self, kernel_size, sigma):
+        gaussian_kernel_x = self._gaussian(kernel_size[0], sigma[0])
+        gaussian_kernel_y = self._gaussian(kernel_size[1], sigma[1])
+
+        return torch.matmul(gaussian_kernel_x.t(), gaussian_kernel_y)  # (kernel_size, 1) * (1, kernel_size)
+
+    @reinit__is_reduced
+    def update(self, output: Sequence[torch.Tensor]) -> None:
+        y_pred, y = output
+        if y_pred.dtype != y.dtype:
+            raise TypeError(
+                "Expected y_pred and y to have the same data type. Got y_pred: {} and y: {}.".format(
+                    y_pred.dtype, y.dtype
+                )
+            )
+
+        if y_pred.shape != y.shape:
+            raise ValueError(
+                "Expected y_pred and y to have the same shape. Got y_pred: {} and y: {}.".format(y_pred.shape, y.shape)
+            )
+
+        if len(y_pred.shape) != 4 or len(y.shape) != 4:
+            raise ValueError(
+                "Expected y_pred and y to have BxCxHxW shape. Got y_pred: {} and y: {}.".format(y_pred.shape, y.shape)
+            )
+
+        C1 = pow(self.k1 * self.data_range, 2)
+        C2 = pow(self.k2 * self.data_range, 2)
+        channel = y_pred.size(1)
+        device = y_pred.device
+        self.kernel = self.kernel.expand(channel, 1, -1, -1).to(device=device)
+
+        mu_pred = F.conv2d(y_pred, self.kernel, groups=channel)
```

Comment:
I'm also wondering whether it won't be more performance interesting to execute a single convolution on a stack of 
`[y_pred, y, y_pred * y_pred, y * y, y_pred * y]` instead of 5 convolutions. Typically, we can imagine that the input image is of size between 256 to 512.

Reply:
Ahh, yeah right, we can do that!

Reply:
I benchmarked that and there is no much gain using that. If you can benchmark that from your side, it would be nice